### PR TITLE
Handle tempita templed cython files

### PR DESCRIPTION
**Pull request by ev-br**
A companion to https://github.com/scikit-image/scikit-image/pull/1923

The code here is basically adopted from https://github.com/scipy/scipy/blob/master/tools/cythonize.py

### CODE REVIEW

**Comment by emmanuelle** [comment]
@ev-br just a naive question: what's the interest of using templates instead of global variables?

**Comment by stefanv** [comment]
@emmanuelle If you take a look at the PR referenced, you'll find a section of code that is generate by a for loop:

```
+ {{for _octant in range(8)}}
+ # octant {{_octant}}:
+    n = 1
+    {{for _j in range(7):}}
+    {{py: _idx = _neighb_idx[_octant][_j]}}
+    if neighbors[{{_idx}}] == 1:
+        n |= {{1 << (7 - _j)}}
+
+    {{endfor}}
+    euler_char += lut[n]
+    {{endfor}}
```

While one could potentially use variables to switch out between copied and slightly modified code sections, this is much cleaner.

File: skimage/_build.py [hunk 1]
```diff
@@ -69,3 +74,21 @@ def _changed(filename):
...
+            from Cython import Tempita as tempita
+        except ImportError:
+            import tempita
+    except ImportError:
+        raise Exception('Building requires Tempita: '
+                        'pip install --user Tempita')
+    from_filename = tempita.Template.from_filename
+    template = from_filename(fromfile, encoding=sys.getdefaultencoding())
+    pyxcontent = template.substitute()
+    assert fromfile.endswith('.pyx.in')
```
Comment (stefanv) [inline skimage/_build.py]: Assert is probably best replaced with a raise ValueError

File: skimage/_build.py [hunk 2]
```diff
@@ -69,3 +74,21 @@ def _changed(filename):
...
+    pyxfile = fromfile[:-len('.pyx.in')] + '.pyx'
```
Comment (stefanv) [inline skimage/_build.py]: os.path.splitext(fromfile)[0] ?

File: skimage/_build.py [hunk 3]
```diff
@@ -69,3 +74,21 @@ def _changed(filename):
...
+def process_tempita_pyx(fromfile):
+    try:
+        try:
+            from Cython import Tempita as tempita
+        except ImportError:
+            import tempita
+    except ImportError:
+        raise Exception('Building requires Tempita: '
+                        'pip install --user Tempita')
+    from_filename = tempita.Template.from_filename
```
Comment (stefanv) [inline skimage/_build.py]: Not really necessary to do this re-assignment, since we don't use it more than once further down

**Comment by stefanv** [comment]
Some minor style comments, but looks OK to me.

**Comment by ev-br** [comment]
@emmanuelle To add to what Stefan said: in this context, a dozen lines of a template helps to avoid either ~100 lines of repetetive boilerplate (check the .pyx file generated from .pyx.in during the build) or a large perf hit due to looping at run time (this is most useful when the loop is over a python data structure). In gh-1923 there are two templated routines, and together they account for about an order of magnitude speedup.

**Comment by emmanuelle** [comment]
@ev-br @stefanv many thanks for the useful explanations, I think I get it now :-). I hadn't understood at first sight that one benefit was to get rid of Python objects and to generate code without such objects.

**Comment by emmanuelle** [comment]
LGTM, and I think for Stéfan as well, therefore merging.

**Pull request merged**